### PR TITLE
PR: Test and close file handles and threads leaks

### LINKED ISSUE #12328: test_mainwindow.py opens too many file descriptors and sockets
Hi,

First, I want to clarify that I am fully aware of the fact, that you do not support running unit tests outside of your own development environments and CIs. However, because I think this might be related to what happens in #12247, I am reporting anyway. Please, don't feel obliged to dig too much into it if you feel this does not apply to the scenarios you are officially supporting.

In my endeavor to stabilize Spyder for the OpenSUSE Linux distro package which I am maintaining, I try to run as many unit tests as possible in order to detect possible package incompatibilites.

Fact is, the execution of `python3 runtests.py --run-slow` always fails with a segfault or breakpoint trap. This happens not always at the same test but consistently a few dozen tests into the test_mainwindow.py suite. *Sometimes* there is a `Too many open files` error message associated with it.

I added a few debug statements to the `main_window` fixture and found out that for each test, the number of open files and sockets increases. Specifically the `pdb_history.sqlite` file is opened many times but not closed at teardown of the test. I suspect this is because of the reuse of the window as introduced in #10444. This could also be the reason why #12247 cannot open any new connections.

I will dig more into this, but any hints would be appreciated.

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #12328

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by impact27** [comment]
I need help to make test pass. I added a detection for threads and files leak in ipyconsole tests, and some tests are leaking threads.

**Comment by bnavigator** [comment]
Hi @impact27, with your recent commits, do you see a teardown at all? Because I was about to write about that before I saw that you added stuff here.

**Comment by bnavigator** [comment]
A teardown of the main_window i mean. Here is the output of this PR applied up to 28651a1, #11704 applied, and some debugging added:

```diff
[runlog.txt](https://github.com/spyder-ide/spyder/files/4488321/runlog.txt)

diff -ur -x __pycache__ spyder-4.1.2o/spyder/app/tests/test_mainwindow.py spyder-4.1.2/spyder/app/tests/test_mainwindow.py
--- spyder-4.1.2o/spyder/app/tests/test_mainwindow.py   2020-04-16 15:37:08.090332131 +0000
+++ spyder-4.1.2/spyder/app/tests/test_mainwindow.py    2020-04-16 15:28:16.002870925 +0000
@@ -25,6 +25,8 @@
 import sys
 import uuid
 import psutil
+from pprint import pprint
+from time import time
 
 # Third party imports
 from IPython.core import release as ipy_release
@@ -151,9 +153,16 @@
 # ---- Fixtures
 # =============================================================================
 
-@pytest.fixture
+@pytest.fixture()
 def main_window(request):
     """Main Window fixture"""
+
+    print(("SETUP", time()))
+    proc = psutil.Process()
+    print(("Num FDs: ", proc.num_fds())) 
+    pprint(proc.open_files())
+    pprint(proc.connections("unix"))
+   
     # Tests assume inline backend
     CONF.set('ipython_console', 'pylab/backend', 0)
 
@@ -206,8 +215,15 @@
         # Reset cwd
         window.explorer.chdir(get_home_dir())
 
+    print(("SETUP DONE", time()))
+    pprint(proc.connections("unix"))
+
     yield window
 
+    print(("TEARDOWN", time()))
+    print(("Num FDs: ", proc.num_fds()))
+    pprint(proc.open_files())
+
     # Print shell content if failed
     if request.node.rep_setup.passed:
         if request.node.rep_call.failed:
@@ -223,6 +239,10 @@
             window.close()
             del main_window.window
 
+    print(("TEARDOWN DONE", time()))
+    print(("Num FDs: ", proc.num_fds()))
+    pprint(proc.open_files())
+
 
 @pytest.fixture(scope="session", autouse=True)
 def cleanup(request):
```
``python3 runtests.py -rP --run-slow -k 'test_default or test_lock_action or test_change_cwd' spyder/app/tests/test_mainwindow.py > ../runlog.txt``
[runlog.txt](https://github.com/spyder-ide/spyder/files/4488323/runlog.txt)

As you can see, there is a `SETUP` and `SETUP DONE` with ever increasing `fd_num()` and more connections open, but no `TEARDOWN`.

**Comment by impact27** [comment]
> 
> 
> Hi @impact27, with your recent commits, do you see a teardown at all? Because I was about to write about that before I saw that you added stuff here.

So the teardown was happening at the beginning of the test (Teardown of the previous test). Here I am doing it in the end, which makes more sense

**Comment by bnavigator** [comment]
So it seems that I don't see the TEARDOWN messages because somehow the output is not captured by pytest anymore. The code after the yield is actually executed after the test ends.

**Comment by bnavigator** [comment]
Hey @impact27, this is great work! Thanks a lot for putting so much effort into this! 

`runtests.py --run-slow test_mainwindow.py` stell gets aborted with core dump on my system, but on a much later point (and still not reproducibly at the same test)

```
spyder/app/tests/test_mainwindow.py::test_runcell[True] PASSED                                   [ 81%]Fatal Python error: Aborted

Thread 0x00007f2ab5db4700 (most recent call first):
  File "/usr/lib64/python3.8/selectors.py", line 468 in select
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 1823 in _run_once
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 570 in run_forever
  File "/usr/lib64/python3.8/site-packages/tornado/platform/asyncio.py", line 149 in start
  File "/usr/lib/python3.8/site-packages/jupyter_client/threaded.py", line 188 in run
  File "/usr/lib64/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib64/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f2ad5016700 (most recent call first):
  File "/usr/lib64/python3.8/threading.py", line 302 in wait
  File "/usr/lib64/python3.8/threading.py", line 558 in wait
  File "/usr/lib/python3.8/site-packages/IPython/core/history.py", line 829 in run
  File "/usr/lib/python3.8/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-24>", line 2 in run
  File "/usr/lib64/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib64/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f2b4eae2700 (most recent call first):
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/plugins/onlinehelp/pydoc_patch.py", line 65 in serve_until_quit
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/plugins/onlinehelp/pydoc_patch.py", line 94 in run
  File "/usr/lib64/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib64/python3.8/threading.py", line 890 in _bootstrap

Current thread 0x00007f2baf838740 (most recent call first):
  File "/usr/lib64/python3.8/site-packages/zmq/sugar/socket.py", line 59 in __init__
  File "/usr/lib64/python3.8/site-packages/zmq/sugar/context.py", line 204 in socket
  File "/usr/lib/python3.8/site-packages/jupyter_client/connect.py", line 547 in _create_connected_socket
  File "/usr/lib/python3.8/site-packages/jupyter_client/connect.py", line 563 in connect_shell
  File "/usr/lib/python3.8/site-packages/jupyter_client/client.py", line 141 in shell_channel
  File "/usr/lib/python3.8/site-packages/jupyter_client/threaded.py", line 238 in start_channels
  File "/usr/lib/python3.8/site-packages/qtconsole/kernel_mixins.py", line 47 in start_channels
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/plugins/ipythonconsole/plugin.py", line 753 in connect_client_to_kernel
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/plugins/ipythonconsole/plugin.py", line 699 in create_new_client
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/plugins/ipythonconsole/plugin.py", line 1022 in close_client
  File "/home/abuild/rpmbuild/BUILD/spyder-4.1.2/spyder/app/tests/test_mainwindow.py", line 239 in main_window
  File "/usr/lib/python3.8/site-packages/_pytest/fixtures.py", line 803 in _teardown_yield_fixture
  File "/usr/lib/python3.8/site-packages/_pytest/fixtures.py", line 852 in finish
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 305 in _callfinalizers
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 315 in _teardown_with_finalization
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 297 in _pop_and_teardown
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 340 in _teardown_towards
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 332 in teardown_exact
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 137 in pytest_runtest_teardown
  File "/usr/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 198 in <lambda>
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 226 in from_call
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 197 in call_runtest_hook
  File "/usr/lib/python3.8/site-packages/flaky/flaky_pytest_plugin.py", line 129 in call_and_report
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 94 in runtestprotocol
  File "/usr/lib/python3.8/site-packages/_pytest/runner.py", line 78 in pytest_runtest_protocol
  File "/usr/lib/python3.8/site-packages/flaky/flaky_pytest_plugin.py", line 92 in pytest_runtest_protocol
  File "/usr/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 271 in pytest_runtestloop
  File "/usr/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 250 in _main
  File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 206 in wrap_session
  File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 243 in pytest_cmdline_main
  File "/usr/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/lib/python3.8/site-packages/_pytest/config/__init__.py", line 82 in main
  File "runtests.py", line 52 in run_pytest
  File "runtests.py", line 69 in main
  File "runtests.py", line 73 in <module>
Aborted (core dumped)
```

More importantly, I can now run most of the `ipythonconsole` tests. Before, those tests were skipped completely in the package. Thanks alot!

**Comment by goanpeca** [comment]
Thanks for working in this. Will take it for a spin today.

**Comment by impact27** [comment]
What you need to know:
 - The teardown of a few tests are slow: We should see if there is a way of closing the extra file/thread quickier.

fast:
```
87.87s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_stderr_file_is_removed_two_kernels
87.59s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_stderr_file_remains_two_kernels
87.51s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_stderr_file_is_removed_one_kernel
45.13s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_no_repeated_tabs_name
45.01s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_tabs_preserve_name_after_move
45.00s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_console_disambiguation
42.00s teardown spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_kernel_crash
```
slow:
```
46.99s teardown spyder/app/tests/test_mainwindow.py::test_runfile_from_project_explorer
46.86s teardown spyder/app/tests/test_mainwindow.py::test_open_notebooks_from_project_explorer
46.20s teardown spyder/app/tests/test_mainwindow.py::test_window_title
```


 -  some kernel stderr files leaks on windows, so I skip the file leak check on windows.

 - `test_custom_layouts` leaks a thread on windows

 - The fast tests on linux seems to hang without reason.

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 1]
```diff
@@ -70,6 +70,11 @@ def __init__(self, *args, **kwargs):
         self._tmp_reading = False
         self._pdb_frame_loc = (None, None)
 
+    def will_close(self, externally_managed):
```
Comment (goanpeca) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:73]: how is this argument used?

Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py:73]: I updated the docstring

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 2]
```diff
@@ -125,7 +125,10 @@ def shutdown(self):
         self.shutdown_thread.start()
 
     def will_close(self, externally_managed):
-        """Close shell"""
+        """
+        Close communication channels with the kernel if shutdown was not called.
+        If the kernel is not extrenally managed, shutdown the kernel as well.
```
Comment (goanpeca) [inline spyder/plugins/ipythonconsole/widgets/shell.py:130]: extrenally -> externally

**Comment by impact27** [comment]
The problems I have with the tests are:

On windows, Some tests will leak a ServerThread. I think the problem is somewhere around `spyder/plugins/onlinehelp/widgets.py` where `start_server` doesn't properly close the self.server.server (This thread only exists in PY3).

On Linux, the slow tests succeed but hang for some reason. My guess is that a process (or a thread) is not closed somehow and this prevents the tests from finishing.

**Comment by ccordoba12** [comment]
> On Linux, the slow tests succeed but hang for some reason. My guess is that a process (or a thread) is not closed somehow and this prevents the tests from finishing.

That's my guess too. But trying to understand which process is the culprit is very hard.

**Comment by impact27** [comment]
> > On Linux, the slow tests succeed but hang for some reason. My guess is that a process (or a thread) is not closed somehow and this prevents the tests from finishing.
> 
> That's my guess too. But trying to understand which process is the culprit is very hard.

I am relatively sure this is related to `test_mainwindow.py`. What would make the most sense is that one kernel process is not close as they are the only subprocesses that are really affected here.

**Comment by bnavigator** [comment]
> > > On Linux, the slow tests succeed but hang for some reason. My guess is that a process (or a thread) is not closed somehow and this prevents the tests from finishing.
> > 
> > 
> > That's my guess too. But trying to understand which process is the culprit is very hard.
> 
> I am relatively sure this is related to `test_mainwindow.py`. What would make the most sense is that one kernel process is not closed as they are the only subprocesses that are really affected here.

Can confirm, as of two days ago, the slow tests in `test_mainwindow.py` are the only suite that still prevents my linux package from building. Still gets hard aborted, presumably because of too many processes open. Did not check with the recent commits you pushed, but the build job is running right now:
https://build.opensuse.org/package/show/home:bnavigator:branches:devel:languages:python:numeric/spyder
https://build.opensuse.org/package/live_build_log/home:bnavigator:branches:devel:languages:python:numeric/spyder/openSUSE_Tumbleweed/x86_64


Some days ago, with a previous version of this PR, I tried running `python runtests.py --run-slow -k "not $i"  $path_to/test_mainwindow.py` with `$i` being every single test in test_mainwindow individually. The abort happens everytime. Running just one of the tests or a smaller set passes everything. So it is not a single test but I suspect it is the fixture setup itself.

**Comment by impact27** [comment]
Closing. If someone wants to try to fix the tests, you can reopen

**Pull request closed without merging**